src/Services/InventoryUpdateService.php

```diff
@@ -37,319 +54,524 @@ class InventoryUpdateService
+      ];
+    }
   }
 
   /**
    * Get the stock buffer value, normalized to 0
    *
-   * @param AbstractConfigHelper $configHelper
-   * @param LoggerContract $loggerContract
    * @return int
    */
-  private static function getNormalizedStockBuffer($configHelper, $loggerContract = null)
+  private function getNormalizedStockBuffer()
   {
     $stockBuffer = null;
-    if (isset($configHelper)) {
-      $stockBuffer = $configHelper->getStockBufferValue();
-    }
+    $stockBuffer = $this->configHelper->getStockBufferValue();
+
 
     if (isset($stockBuffer)) {
       if ($stockBuffer >= 0) {
         return $stockBuffer;
       }
 
       // invalid value for buffer
-      if (isset($loggerContract)) {
-        $loggerContract->warning(
-          TranslationHelper::getLoggerKey(self::LOG_KEY_INVALID_STOCK_BUFFER),
-          [
-            'additionalInfo' => ['stockBuffer' => $stockBuffer],
-            'method' => __METHOD__
-          ]
-        );
-      }
+      $this->logger->warning(
+        TranslationHelper::getLoggerKey(self::LOG_KEY_INVALID_STOCK_BUFFER),
+        [
+          'additionalInfo' => [
+            'stockBuffer' => json_encode($stockBuffer)
+          ],
+          'method' => __METHOD__
+        ]
+      );
     }
 
     return 0;
   }
+
+  /**
+   * Check if the state of the Inventory Service has been "running" for more than the maximum allotted time
+   * This functionality was extracted from the old UpdateFullInventoryStatusCron
+   * @param bool $full check full inventory service instead of partial
+   * @return bool
+   */
+  private function serviceHasBeenRunningTooLong(bool $full): bool
+  {
+
+    if (!$this->statusService->isInventoryRunning($full)) {
+      return false;
+    }
+
+    $maxTime = $full ? self::MAX_INVENTORY_TIME_FULL : self::MAX_INVENTORY_TIME_PARTIAL;
+    $logKey = $full ? self::LOG_KEY_LONG_RUN_FULL : self::LOG_KEY_LONG_RUN_PARTIAL;
+
+    $lastStart = $this->statusService->getLastAttemptTime($full);
+    if (isset($lastStart) && !empty($lastStart) && (time() - strtotime($lastStart)) > $maxTime) {
+
+      $this->logger->warning(TranslationHelper::getLoggerKey($logKey), [
+        'additionalInfo' => ['startedOn' => $lastStart, 'maximumTime' => $maxTime],
+        'method' => __METHOD__
+      ]);
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Get the php time (in seconds) for the start of a partial sync
+   *
+   * @param ExternalLogs $externalLogs
+   *
+   * @return int
+   */
+  private function getStartOfDeltaSyncWindow(ExternalLogs $externalLogs = null): int
+  {
+    $windowStart = 0;
+
+    $lastWindowEnd = $this->statusService->getLastSuccessfulAttemptTime(false);
+    $lastGoodFullStart = $this->statusService->getLastSuccessfulAttemptTime(true);
+
+    if (isset($lastWindowEnd) && !empty($lastWindowEnd)) {
+      // new window should be directly after the previous window
+      $windowStart = strtotime($lastWindowEnd);
```

Comment:
Hm, definitely something to think about. We did see (lack of timezone) issues in the UI and there is likely to be something similar going on here.
That effort would increase the scope of this PR though, because we'd need to tell the UI how to interpret/display the values. Right now it just does a straight-through display of what's in storage.

Reply:
Actually I think the timestamps we have in storage _do_ have timezones on them...

Reply:
yeah @jiafanwu the values in our DB are like this one: `2020-08-31 01:09:15.000000 +02:00` so, we should not see timezone issues.
I agree that the conversions are not great, but I'd rather be doing conversions here on the back-end than having to do them on the front-end.